src/inputdevice.cpp

```diff
@@ -4993,6 +4994,65 @@ static uae_u64 isqual (int evt)
 	return ID_FLAG_QUALIFIER1 << (num * 2);
 }
 
+// Pass the joystick state (joybutton and joydir) to kvbd subsystem and clear them for simulator.
+static void handle_vkbd()
+{
+	if (!vkbd_is_active())
+	{
+		return;
+	}
+
+	for(int joy = 0; joy < MAX_JPORTS; ++joy)
+	{
+		oleft[joy] = 0;
+		oright[joy] = 0;
+		otop[joy] = 0;
+		obot[joy] = 0;
+		horizclear[joy] = 0;
+		vertclear[joy] = 0;
+		mouse_x[joy] = 0;
+		mouse_y[joy] = 0;
+	}
+
+
+	int vkbd_state = 0;
+	for(int joy = 0; joy < MAX_JPORTS; ++joy)
+	{
+		int mask_button = 1 << JOYBUTTON_1;
```

Comment:
This works for Default, Joystick and Gamepad modes. But not if the controller is set to CD32 mode - then the fire button / button 0 does not have any effect. You can bring up the VKBD with the custom control, navigate to the buttons, but you cannot press any. :)

Reply:
I will take a look at these things next week.

Reply:
@midwan I don't know anything about CD32. Is this when you select in 'CD32 pad' in 'Input' screen left or 'Remap' button? What is the equivalent of fire button on CD32 (I figured out that JOYBUTTON_1 corresponds to fire button on joystick by looking at custom controls. For CD32 I see a lot of things in custom controls, but what corresponds to fire button?)

Reply:
@erikbernsen 
The `mode` of the controller is what determines the mapping that will be used. CD32 pads have more buttons supported, so they are only activated if you select that mode. The other modes work fine, but when CD32 mode is selected, the `RED` button is the first "Fire" button, instead of JOYBUTTON_1.

Reply:
@erikbernsen 
With the latest changes the mouse problem is resolved, good job! :)
For the CD32 button, you can also take a look at `inputdevice.cpp` line 3655, function `handle_joystick_buttons`:

```
for (i = 0; i < 2; i++) {
		int mask = 0x40 << i;
		if (cd32_pad_enabled[i]) {
			but |= mask;
			if (!cd32padmode(i)) {
				if (getbuttonstate (i, JOYBUTTON_CD32_RED) || getbuttonstate (i, JOYBUTTON_1))
					but &= ~mask;
				// always zero if output=1 and data=0
				if ((dra & mask) && !(pra & mask)) {
					but &= ~mask;
				}
			}
```

I guess it would be OK to only accept input from the first 2 joysticks, not the ones in the (virtual) parallel port, therefore the loop could iterate from 0 to < 2 like in the example above.
Then in the same class, you have `cd32_pad_enabled[i]` which can tell you if the mode is indeed CD32 or not. And finally, as you can see in the example, you can check for both `JOYBUTTON_1` and `JOYBUTTON_CD32_RED` to determine if "Fire" was pressed.

Reply:
OK, this is something I can fix myself also, so to speed up the process I'll merge it as-is for now, and fix it afterwards. :)